huntsman/camera/pyro.py

```diff
@@ -275,55 +261,44 @@ def autofocus(self,
 
+        Raises:
+            ValueError: If invalid values are passed for any of the focus parameters.
+        """
         if self.focuser is None:
             msg = "Camera must have a focuser for autofocus!"
             self.logger.error(msg)
             raise AttributeError(msg)
 
-        assert self.focuser.is_connected, \
-            self.logger.error("Focuser must be connected for autofocus.")
-
-        # Compile aruments into a dictionary
-        autofocus_kwargs = {'seconds': seconds,
-                            'focus_range': focus_range,
-                            'focus_step': focus_step,
-                            'keep_files': keep_files,
-                            'take_dark': take_dark,
-                            'thumbnail_size': thumbnail_size,
-                            'merit_function': merit_function,
-                            'merit_function_kwargs': merit_function_kwargs,
-                            'mask_dilations': mask_dilations,
-                            'coarse': coarse,
-                            'make_plots': make_plots}
-        autofocus_kwargs.update(kwargs)
-
-        # Make sure proxy is in async mode
-        Pyro4.asyncproxy(self._proxy, asynchronous=True)
-
-        # Start autofocus
-        autofocus_result = {}
-        self.logger.debug('Starting autofocus on {}'.format(self.name))
-        # Remote method call to start the autofocus
-        autofocus_result = self._proxy.autofocus(*args, **autofocus_kwargs)
-
-        # Start a thread that will set an event once autofocus has completed
-        autofocus_event = Event()
-        autofocus_thread = Thread(target=self._async_wait,
-                                  args=(autofocus_result, 'autofocus', autofocus_event, timeout))
-        autofocus_thread.start()
+        self.logger.debug(f'Starting autofocus on {self}.')
+
+        # Remote method call to start the exposure
+        self._proxy.autofocus(*args, **kwargs)
 
+        # Proxy for remote _autofocus_event
+        self._autofocus_event = RemoteEvent(self._proxy, event_type="focuser")
+
+        # In general it's very complicated to work out how long an autofocus should take
+        # because parameters can be set here or come from remote config. For not just make
+        # it 5 minutes.
+        max_wait = 300
         if blocking:
-            autofocus_event.wait()
+            success = self._autofocus_event.wait(timeout=max_wait)
+            if not success:
+                self._timeout_response("autofocus")
+        else:
+            # If the remote autofocus fails after starting in such a way that the event doesn't
+            # doesn't get set then calling code could wait forever. Have a local timeout thread
+            # to be safe.
+            timeout_thread = Timer(interval=max_wait,
+                                   function=self._timeout_response,
+                                   args=("autofocus",))
+            timeout_thread.start()
```

Comment:
This is almost a repeat of (new) lines 210-218, might be worth wrapping up into a helper method. Might no. :laughing:

Reply:
Dangling more opportunities for reducing line count in front of me... OK, done.